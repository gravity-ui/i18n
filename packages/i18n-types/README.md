# Общие типы i18n
