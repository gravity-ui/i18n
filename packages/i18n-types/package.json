{
  "name": "@gravity-ui/i18n-types",
  "version": "1.0.0",
  "license": "UNLICENSED",
  "main": "dist/cjs/index.js",
  "module": "dist/esm/index.js",
  "types": "dist/esm/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "pnpm run build:clean && pnpm run build:compile",
    "build:compile": "tsc -p tsconfig.json && tsc -p tsconfig.cjs.json",
    "build:clean": "rm -rf dist",
    "lint": "eslint .",
    "lint:fix": "eslint . --fix",
    "test": "jest",
    "typecheck": "tsc --noEmit"
  },
  "keywords": [
    "i18n"
  ],
  "devDependencies": {
    "eslint": "^9.15.0",
    "eslint-config": "workspace:*",
    "typescript": "^5.6.3",
    "@types/jest": "^29.5.14",
    "jest": "29.7.0",
    "ts-jest": "29.1.1"
  },
  "engines": {
    "node": ">= 18"
  }
}
