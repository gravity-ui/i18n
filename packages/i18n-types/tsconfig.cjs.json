{
    "extends": "../../tsconfig.base.json",
    "compilerOptions": {
        "outDir": "./dist/cjs",
        "module": "CommonJS"
    },
    "include": ["src/*.ts", "src/**/*.ts"],
    "exclude": ["**/__tests__/**"]
}
