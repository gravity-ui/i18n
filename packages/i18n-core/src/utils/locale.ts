import {MissingDataError} from '@formatjs/intl';

import type {ResolvedIntlConfig} from '../types';

import {getFallbackLocalesList} from './fallbacks';
import {TECH_LOCALE} from '@gravity-ui/i18n-types';

export const validateLocale = <AvailableLocale extends string>(
    locale: string,
    config: Pick<
        ResolvedIntlConfig<never, AvailableLocale>,
        | 'allowedLocales'
        | 'fallbackLocales'
        | 'onError'
        | 'onWarn'
        | 'disableUseLocaleLangAsFallback'
        | 'defaultFallback'
    >,
) => {
    const {allowedLocales, onError, onWarn} = config;

    if (locale === TECH_LOCALE) {
        return true;
    }

    try {
        if (!Intl.NumberFormat.supportedLocalesOf(locale).length) {
            onError?.(
                new MissingDataError(
                    `Missing locale data for locale: "${locale}" in Intl.DateTimeFormat.`, // Using previous locale as fallback.
                ),
            );
            return false;
        }
    } catch {
        onError?.(
            new MissingDataError(
                `Missing locale data for locale: "${locale}" in Intl.DateTimeFormat.`, // Using previous locale as fallback.
            ),
        );
        return false;
    }

    if (!allowedLocales.includes(locale as AvailableLocale)) {
        const fallbackLocalesForLocale = getFallbackLocalesList(locale, config);

        if (fallbackLocalesForLocale.length === 0) {
            onError?.(
                new MissingDataError(
                    `Missing locale data for locale: "${locale}" in config.allowedLocales and in config.fallbackLocales.`, // Using previous locale as fallback.
                ),
            );
            return false;
        }

        onWarn?.(`Missing locale data for locale: "${locale}" in config.`); // Using [${fallbackLocalesForLocale}] locales as fallback.
    }

    return true;
};
