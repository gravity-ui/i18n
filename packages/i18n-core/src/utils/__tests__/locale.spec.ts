import {TECH_LOCALE} from '@gravity-ui/i18n-types';
import {validateLocale} from '../locale';

describe('locale', () => {
    it('should be valid if it is one of allowed locales', () => {
        let error;

        expect(
            validateLocale('en', {
                allowedLocales: ['en'],
                fallbackLocales: {},
                onError: (e) => {
                    error = e;
                },
                onWarn: () => {},
                disableUseLocaleLangAsFallback: true,
                defaultFallback: 'empty-string',
            }),
        ).toBe(true);

        expect(error).toEqual(undefined);
    });

    it('should be valid if have fallback locales', () => {
        let error;

        expect(
            validateLocale('en-kz', {
                allowedLocales: ['en'],
                fallbackLocales: {
                    // @ts-ignore
                    'en-kz': 'en',
                },
                onError: (e) => {
                    error = e;
                },
                onWarn: () => {},
                disableUseLocaleLangAsFallback: true,
                defaultFallback: 'empty-string',
            }),
        ).toBe(true);

        expect(error).toEqual(undefined);
    });

    it('should not be valid if it is unknown locale for intl', () => {
        let errorA;

        expect(
            validateLocale('unknown', {
                allowedLocales: ['en'],
                fallbackLocales: {
                    // @ts-ignore
                    'en-kz': 'en',
                },
                onError: (e) => {
                    errorA = e;
                },
                onWarn: () => {},
                disableUseLocaleLangAsFallback: true,
                defaultFallback: 'empty-string',
            }),
        ).toBe(false);

        expect(errorA).toMatchSnapshot('not valid locale error');

        let errorB;

        expect(
            validateLocale('someUnknownLocale', {
                allowedLocales: ['en'],
                fallbackLocales: {
                    // @ts-ignore
                    'en-kz': 'en',
                },
                onError: (e) => {
                    errorB = e;
                },
                onWarn: () => {},
                disableUseLocaleLangAsFallback: true,
                defaultFallback: 'empty-string',
            }),
        ).toBe(false);

        expect(errorB).toMatchSnapshot('not valid locale for Intl error');
    });

    it('should be valid if it is tech locale', () => {
        expect(
            validateLocale(TECH_LOCALE, {
                allowedLocales: ['en'],
                fallbackLocales: {},
                onError: () => {},
                onWarn: () => {},
                disableUseLocaleLangAsFallback: true,
                defaultFallback: 'empty-string',
            }),
        ).toBe(true);
    });
});
