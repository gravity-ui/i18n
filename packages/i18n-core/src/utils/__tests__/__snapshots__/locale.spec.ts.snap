// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`locale should not be valid if it is unknown locale for intl: not valid locale error 1`] = `
[Error: [@formatjs/intl Error MISSING_DATA] Missing locale data for locale: "unknown" in Intl.DateTimeFormat.
]
`;

exports[`locale should not be valid if it is unknown locale for intl: not valid locale for Intl error 1`] = `
[Error: [@formatjs/intl Error MISSING_DATA] Missing locale data for locale: "someUnknownLocale" in Intl.DateTimeFormat.
]
`;
